### PR TITLE
Enhancement/integration tests

### LINKED ISSUE #128: Smoke tests to cover all providers
## Objective
To have integration/smoke tests
- Across all cloud providers
- Kubernetes versions
- To avoid breakage while adding new features/fixes

## Current state of smoke tests
We have integration tests written on bash (only for AWS) where we run a set of smoke tests like
1. Create and delete single machine object
1. Create a machine-deployment with 3 replicas
1. Scale it down to 2
1. Scale it up to 4
1. Rolling update of machines from v1 to v2

## Possible solutions moving forward
We have two options moving forward
1. **Improve smoke tests** to check for all providers and K8s versions by provisioning clusters statically (clusters that are pre-created) /dynamically(on-demand)
1. **Mock API calls** for different cloud providers to mock providers while running integration tests

### Improve smoke tests
Things to consider
- How many MCM would be running?
- Where would MCM be running?
- Different poroviders supporting different k8s versions?
- Control-target only for one provider?
- Clean up - Run a stable controller and remove an partially created resources.

### 2. Mock API calls
- AWS: mocking available
- Openstack: Two ways, but mocking avaiable
- GCP: Not avaiable
- Azure: Issue/PR was opened and closed

### PR DESCRIPTION
**Pull request by prashanth26**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #128 

**Special notes for your reviewer**:
This PR adds integration test support for all providers and different Kubernetes versions

**Release note**:

```improvement user
- Enhanced integration tests to support AWS and GCP. 
- Adding support in integration tests for a new provider is simpler and modular now.
```

### CODE REVIEW

**Comment by prashanth26** [comment]
Raising this PR to start a review, in the meanwhile I am working on fixing the issues with Openstack

File: .ci/integration-tests [hunk 1]
```diff
@@ -56,18 +90,23 @@ function setup_environment() {
...
-                --target-kubeconfig=$target_kubeconfig \
-                --namespace=$control_namespace \
-                --safety-up=2 \
-                --safety-down=1 \
-                --machine-drain-timeout=5 \
-                --machine-health-timeout=10 \
-                --machine-set-scale-timeout=20 \
-                --v=2 > mcm.out 2>&1 &
+        --control-kubeconfig=dev/control_kubeconfig_${provider}.yaml \
+        --target-kubeconfig=dev/control_kubeconfig_${provider}.yaml \
```
Comment (PadmaB) [inline .ci/integration-tests]: Should this be target_kubeconfig_${provider}.yaml ?

Comment (prashanth26) [reply .ci/integration-tests]: Yes, you are right. I am wondering how the integration tests succeeded then ❓

File: .ci/integration-tests [hunk 2]
```diff
@@ -183,58 +235,158 @@ function tc_machine_deployment() {
...
+        ret=$?
+        [[ $ret -ne 0 ]]
+    then
+        # Delete the machine-deployment
+        hf_object_delete $provider/md.yaml
+        printf "\n\tWaiting 1200s for machine-deployment to be deleted"
+        hf_wait_on "hf_num_of_objects" machdeploy 0 1200 $5
+    fi
+
+    if
```
Comment (hardikdr) [inline .ci/integration-tests]: Should we add MachineSet as well?

Comment (prashanth26) [reply .ci/integration-tests]: Added machine-set in the clean-up process as suggested.

**Comment by bergerx** [comment]
Just our of curiosity, what was the issue that caused to drop Azure support in 2ad546c?

**Comment by prashanth26** [comment]
Hi @bergerx ,
                       Sorry about that, its just about integration tests. The support for Azure still exists. Its just that the Azure clusters for running integration tests haven't been setup yet. We shall be adding integration tests to test for Azure and Openstack very soon.

Thanks & Regards,
Prashanth

**Pull request merged**